### ISSUE TITLE
Error running function with NodePort router

### ISSUE THREAD

**Issue by jacobpiggin** [fission/fission#1631]
Kubernetes v1.18
Fission v1.9.0

Kubeadm on a single master node (CentOS8 VM) and WeaveNet addon. 

**Describe the bug**
Set up fission with serviceType and routerServiceType = NodePort as my kube cluster doesn't have LoadBalancer support.
When running fission fn test --name hello I get:

` Error: error executing HTTP request: Get http://<ip address>:<random port e.g. 45153>/fission-function/hello: context deadline exceeded`

I also get:
`Warning: The environment variable FISSION_ROUTER is no longer supported for this command` 
Since the port changes every time I run the command, I disabled the firewall temporarily to see if this fixed the issue but it did not.

When trying to curl `<ip address>:31314/hello` I only get "error sending request to function" returned. The function logs don't show anything.

The router service has the type NodePort and port 31314 displayed. I've tested using NodePort outside of Fission following [this example](https://kubernetes.io/docs/tasks/access-application-cluster/service-access-application-cluster/) and it works correctly so I don't think it's an issue with my Kubernetes installation.

**To Reproduce**
Install Fission using Helm, by running the command:
`helm install --namespace $FISSION_NAMESPACE --name-template fission \
    --set serviceType=NodePort,routerServiceType=NodePort \
    https://github.com/fission/fission/releases/download/1.9.0/fission-all-1.9.0.tgz`

Try to run the hello nodejs example.
I'm running a CentOS8 VM.

**Expected result**
Return response "Hello, world!"

**Actual result**
```
Warning: The environment variable FISSION_ROUTER is no longer supported for this command

Options:
  ...
Global Options:
...
Usage:
  fission function test [options]

"Error: error executing HTTP request: Get http://<ip address>:45153/fission-function/hello: context deadline exceeded"
```

Thanks.

**Comment by vishal-biyani** [comment]
Hi @jacobpiggin Thanks for the details. The `fission fn test` command uses Kubectl based port forwarding for testing functions. Can you please verify two things for me:

1) Run the tets command with higher verbosity and share the output please:

`fission fn test --name fn_name --verbosity=3`
 
2) Check if you can port-forward the controller port (By replacing the name of actual controller pod name)

$kubectl port-forward controller-7cdbddf79f-sk6cl 8888

And once that works, in another console you should be able to curl to it:

```
$ curl http://127.0.0.1:8888/

{"Build":{"GitCommit":"8a4d2c8ecc821e896d87f074660685565156aff1","BuildDate":"2020-05-10T16:11:44Z","Version":"1.9.0"},"ServerTime":{"Timezone":"UTC","CurrentTime":"2020-06-05T06:17:27.560428042Z"}}%
```

**Comment by jacobpiggin** [comment]
Thanks for the response @vishal-biyani I re-ran `fission fn test --name fn_name --verbosity=3` with the nodejs example and got the response expected. 

Then I tried to run the HelloWorld java example and had similar issues so the following output corresponds to the Java example.

1. `fission fn test --name java-hello --verbosity=3` Result:
```
Setting up port forward to application=fission-api in namespace
Waiting for local port 34143
Starting port forward from local port 34143
Waiting for port forward 34143 to start...
Using kubeconfig from environment "/root/.kube/config"
Connected to Kubernetes API
Connecting to port 8888 on pod fission/fission
Starting port forwarder
Forwarding from 127.0.0.1:34143 -> 8888
Forwarding from [::1]:34143 -> 8888
Port forward from local port 34143 started
Warning: The environment variable FISSION_ROUTER is no longer supported for this command
Setting up port forward to application=fission-router in namespace
Handling connection for 34143
Waiting for local port 34725
Starting port forward from local port 34725
Waiting for port forward 34725 to start...
Using kubeconfig from environment "/root/.kube/config"
Connected to Kubernetes API
Connecting to port 8888 on pod fission/fission
Starting port forwarder
Forwarding from 127.0.0.1:34725 -> 8888
Forwarding from [::1]:34725 -> 8888
Port forward from local port 34725 started
Function test url: http://127.0.0.1:34725/fission-function/java-hello
Handling connection for 34725
Handling connection for 34725
Error: Error calling function java-hello: 502; Please try again or fix the error: error sending request to function
Try to get pod logs from controller 'logs/java-hello'
Handling connection for 34143
Error: Error getting function logs from controller: error getting logs from controller, status code: '500'. try to get logs from log database.
Setting up port forward to application=fission-api in namespace
Waiting for local port 39945
Waiting for port forward 39945 to start...
Using kubeconfig from environment "/root/.kube/config"
Connected to Kubernetes API
Connecting to port 8888 on pod fission/fission
Starting port forwarder
Forwarding from 127.0.0.1:39945 -> 8888
Forwarding from [::1]:39945 -> 8888
Port forward from local port 39945 started
Handling connection for 39945
Handling connection for 39945

Options:
...
Global Options:
...
Usage:
...

Error: error getting function response
```

Then I ran `kubectl -n fission logs controller-86f464c5bf-26zp7`:
```
{"level":"info","ts":1591347428.1476195,"caller":"fission-bundle/main.go:123","msg":"skipping trace exporter registration"}
{"level":"info","ts":1591347428.356129,"logger":"controller","caller":"canaryconfigmgr/canaryConfigMgr.go:79","msg":"try to start canary config manager with prometheus service url","prometheus":"http://fission-promotheus-server.fission"}
{"level":"info","ts":1591347428.356416,"logger":"controller.canary_config_manager","caller":"canaryconfigmgr/canaryConfigMgr.go:143","msg":"started canary configmgr controller"}
{"level":"info","ts":1591347428.3564255,"logger":"controller","caller":"controller/config.go:41","msg","started canary config manager"}
{"level":"info","ts":1591347428.3570602,"logger":"controller.api","caller":"controller/api.go:264","msg":"server started","port":8888}
{"level":"error","ts":1591348003.3439143,"logger":"controller.api","caller":"controller/api.go:121","msg":"no active pods found","code":500}
{"level":"error","ts":1591348094.2232375,"logger":"controller.api","caller":"controller/api.go:121","msg":"no active pods found","code":500}

```

2. `kubectl -n fission port-forward controller-86f464c5bf-26zp7 8888` for both nodejs and java Hello World examples with their respective url:
```
$ curl 127.0.0.1:8888/hello
404 page not found
$ curl 127.0.0.1:8888/java-hello
404 page not found
```

When I run `kubectl -n fission get all` I can see some deployments aren't running/available:
```
deployment.apps/fission-promotheus-alertmanager 0/1 READY 1 UP-TO-DATE 0 AVAILABLE 
deployment.apps/fission-promotheus-server 0/1 READY 1 UP-TO-DATE 0 AVAILABLE
deployment.apps/storagesvc 0/1 READY 1 UP-TO-DATE 0 AVAILABLE
```

The services relating to these are listed, but these pods aren't running:
```
pod/fission-promotheus-alertmanager-5844d99569-f2khn 0/2 READY PENDING 0 RESTARTS
pod/fission-promotheus-server-5f947998b9-q6wlb 0/2 READY PENDING 0 RESTARTS
pod/storagesvc-8484ff84db-gnfrb 0/1 READY PENDING 0 RESTARTS
```
All other pods,services,deployments,etc are running and ready except these ones.

**Comment by vishal-biyani** [comment]
Hi @jacobpiggin So the issue I think is that the storage for Storage service, Prometheus and Alert manager need a volume for storing their data and hence it looks like they are not running. 

If you describe those pods, you should see a mounted volume (This will be visible only once they are running properly.

```
  Mounts:
      /data from storage-volume (rw)
```

You will also see PVCs bound:

```
$ kubectl get pvc
storage-service                Bound    pvc-f5a31219-733f-412c-9fa7-e4ba054b1ab2   2Gi        RWO            standard       129m
fission-prometheus-alertmanager   Bound    pvc-f5a31219-733f-412c-9fa7-e4ba054b1ea8   2Gi        RWO            standard       129m
fission-prometheus-server         Bound    pvc-a6a2b4c8-97de-4d32-a688-b884b2aaa63f   8Gi        RWO            standard       129m
```

So you will have to either use a dynamically provisioned volume - or use hostpath as a volume (Using storageClass in values.yaml for Storage Service. Similarly you will have to override for Prometheus too - but that is not needed for running of function)

Last note - your nodejs functions function works because code must be smaller than 1MB and hence it stores the code on the function object itself. If code is bigger than 1MB - then you need to store on storage service and use a volume - hence the failure with java and not with NodeJS function!

**Comment by jacobpiggin** [comment]
@vishal-biyani Yes that seems to be the problem. 
`kubectl describe storagesvc` output includes two mounts:
`/fission from fission-storage (rw)`
`/var/run/secrets/kubernetes.io/serviceaccount from fission-svc-token (ro)`
but it also has warnings:
`Warning  Failed Scheduling  <unknown>  default-scheduler  running "VolumeBinding" filter plugin for pod "storagesvc": pod has unbound immediate PersistentVolumeClaims`

`kubectl get pvc` shows all three PVCs in a pending status.

Thank you, that makes sense! I'll try out those suggestions.

**Comment by vishal-biyani** [comment]
I am closing this for now - but please feel free to reopen if related or a new one if you face other issue.
